### PR TITLE
JSON Support for parsing NaN, Infinity and -Infinity

### LINKED ISSUE #12213: Support Infinity, -Infinity and NaN in read_json
While these special values are not strictly standard conform, most implementations do allow or use them.

For example Google's GSON Library or python's json module in the standard library.

### PR DESCRIPTION
**Pull request by WillAyd**
- [X] closes #12213
- [X] tests added / passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

Not sure if we totally want this but figured I'd give it a shot any way. `simplejson` and the stdlib by comparison would decode these values

### CODE REVIEW

**Comment by WillAyd** [comment]
Ran benchmarks just to be safe and no changes appeared

File: pandas/_libs/src/ujson/python/JSONtoObj.c [hunk 1]
```diff
@@ -502,10 +506,11 @@ PyObject *JSONToObj(PyObject *self, PyObject *args, PyObject *kwargs) {
...
         Object_newTrue,   Object_newFalse,      Object_newNull,
-        Object_newObject, Object_endObject,     Object_newArray,
-        Object_endArray,  Object_newInteger,    Object_newLong,
-        Object_newDouble, Object_releaseObject, PyObject_Malloc,
-        PyObject_Free,    PyObject_Realloc};
+	Object_newPosInf, Object_newNegInf,     Object_newObject,
+	Object_endObject,     Object_newArray,  Object_endArray,
+	Object_newInteger,    Object_newLong,   Object_newDouble,
+	Object_releaseObject, PyObject_Malloc, PyObject_Free,
+	PyObject_Realloc};
```
Comment (jbrockmendel) [inline pandas/_libs/src/ujson/python/JSONtoObj.c]: does the alignment here matter?

Comment (WillAyd) [reply pandas/_libs/src/ujson/python/JSONtoObj.c]: Purely stylistic but I'll fix it up

File: pandas/tests/io/json/test_pandas.py [hunk 2]
```diff
@@ -1597,3 +1597,10 @@ def test_json_indent_all_orients(self, orient, expected):
     def test_json_negative_indent_raises(self):
         with pytest.raises(ValueError, match="must be a nonnegative integer"):
             pd.DataFrame().to_json(indent=-1)
+
+    def test_emca_262_nan_inf_support(self):
```
Comment (jreback) [inline pandas/tests/io/json/test_pandas.py]: can you add the strings, 'NaN', '-Infinity', and 'Infinity' to be sure we are not over parsing. also do we parse more common 'Inf' and '-Inf' ?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/io/json/test_pandas.py (hunk 2 above)
Comment (WillAyd) [reply pandas/tests/io/json/test_pandas.py]: Good idea. So note that I also had to add an "a" to this as the parsing worked correctly, but a subsequent `astype("float64")` that the JSON code does would turn the strings into their float representation.

W.r.t "Inf" and "-Inf" the upside to not doing them is that they aren't part of [EMCA 262](https://www.ecma-international.org/ecma-262/10.0/index.html#Title) so we could point to that document if we don't include. The one gap from EMCA 262 that we don't have is "+Infinity" which I am happy to add here

**Review by jreback — APPROVED** [review]
lgtm. needs a rebase, merge on green.

**Comment by WillAyd** [comment]
Great thanks Jeff

**Pull request merged**